pkg/operator/controller/dns_status_test.go

```diff
@@ -0,0 +1,379 @@
+			availType:      operatorv1.OperatorStatusTypeAvailable,
+			degradeStatus:  operatorv1.ConditionTrue,
+			progressStatus: operatorv1.ConditionFalse,
+			availStatus:    operatorv1.ConditionFalse,
+		},
+		{
+			description:    "no cluster ip, 0/2 pods available",
+			clusterIP:      "",
+			avail:          0,
+			desire:         2,
+			degradeType:    operatorv1.OperatorStatusTypeDegraded,
+			progressType:   operatorv1.OperatorStatusTypeProgressing,
+			availType:      operatorv1.OperatorStatusTypeAvailable,
+			degradeStatus:  operatorv1.ConditionTrue,
+			progressStatus: operatorv1.ConditionTrue,
+			availStatus:    operatorv1.ConditionFalse,
+		},
+		{
+			description:    "no cluster ip, 1/2 pods available",
+			clusterIP:      "",
+			avail:          1,
+			desire:         2,
+			degradeType:    operatorv1.OperatorStatusTypeDegraded,
+			progressType:   operatorv1.OperatorStatusTypeProgressing,
+			availType:      operatorv1.OperatorStatusTypeAvailable,
+			degradeStatus:  operatorv1.ConditionTrue,
+			progressStatus: operatorv1.ConditionTrue,
+			availStatus:    operatorv1.ConditionTrue,
+		},
+		{
+			description:    "no cluster ip, 2/2 pods available",
+			clusterIP:      "",
+			avail:          2,
+			desire:         2,
+			degradeType:    operatorv1.OperatorStatusTypeDegraded,
+			progressType:   operatorv1.OperatorStatusTypeProgressing,
+			availType:      operatorv1.OperatorStatusTypeAvailable,
+			degradeStatus:  operatorv1.ConditionTrue,
+			progressStatus: operatorv1.ConditionFalse,
+			availStatus:    operatorv1.ConditionTrue,
+		},
+		{
+			description:    "daemonset pod available on 0/0 nodes",
+			clusterIP:      "1.2.3.4",
+			avail:          0,
+			desire:         0,
+			degradeType:    operatorv1.OperatorStatusTypeDegraded,
+			progressType:   operatorv1.OperatorStatusTypeProgressing,
+			availType:      operatorv1.OperatorStatusTypeAvailable,
+			degradeStatus:  operatorv1.ConditionFalse,
+			progressStatus: operatorv1.ConditionFalse,
+			availStatus:    operatorv1.ConditionFalse,
+		},
+		{
+			description:    "daemonset pod available on 0/2 nodes",
+			clusterIP:      "1.2.3.4",
+			avail:          0,
+			desire:         2,
+			degradeType:    operatorv1.OperatorStatusTypeDegraded,
+			progressType:   operatorv1.OperatorStatusTypeProgressing,
+			availType:      operatorv1.OperatorStatusTypeAvailable,
+			degradeStatus:  operatorv1.ConditionFalse,
+			progressStatus: operatorv1.ConditionTrue,
+			availStatus:    operatorv1.ConditionFalse,
+		},
+		{
+			description:    "daemonset pod available on 1/2 nodes",
+			clusterIP:      "1.2.3.4",
+			avail:          1,
+			desire:         2,
+			degradeType:    operatorv1.OperatorStatusTypeDegraded,
+			progressType:   operatorv1.OperatorStatusTypeProgressing,
+			availType:      operatorv1.OperatorStatusTypeAvailable,
+			degradeStatus:  operatorv1.ConditionFalse,
+			progressStatus: operatorv1.ConditionTrue,
+			availStatus:    operatorv1.ConditionTrue,
```

Comment:
Would it make sense to say that the service is degraded if it is available on some but not all nodes? It is still available, but for pods on a node where the daemonset is not running, DNS requests will be routed via the service proxy to another node, so performance and possibly reliability may be slightly degraded.

Reply:
Yes, I will update the condition to consider the ClusterIP status.